android_sample/app/src/main/jni/amber_script.cpp

```diff
@@ -0,0 +1,144 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "amber_script.h"
+
+#include "src/make_unique.h"
+
+namespace amber {
+namespace android {
+namespace {
+
+const std::string kAmberDir("amber/");
+const std::string kAmberScriptExtension(".amber");
+const std::string kShaderNameSignature(".vk_shader_");
+const std::string kShaderExtension(".spv");
+
+bool IsEndedWith(const std::string& path, const std::string& end) {
+  const size_t path_size = path.size();
+  const size_t end_size = end.size();
+  if (path_size < end_size)
+    return false;
+
+  return path.compare(path_size - end_size, end_size, end) == 0;
+}
+
+bool IsStartedWith(const std::string& path, const std::string& start) {
+  const size_t path_size = path.size();
+  const size_t start_size = start.size();
+  if (path_size < start_size)
+    return false;
+
+  return path.compare(0, start_size, start) == 0;
+}
+
+std::string GetShaderID(const std::string& shader_name) {
+  size_t spv_extension_pos = shader_name.find_last_of('.');
+  size_t shader_id_pos =
+      shader_name.find_last_of('.', spv_extension_pos - 1UL) + 1UL;
+  return shader_name.substr(shader_id_pos, spv_extension_pos - shader_id_pos);
+}
+
+}  // namespace
+
+AmberScriptLoader::AmberScriptLoader(android_app* app) : app_context_(app) {}
+
+AmberScriptLoader::~AmberScriptLoader() = default;
+
+Result AmberScriptLoader::LoadAllScriptsFromAsset() {
+  FindAllScripts();
+  if (script_info_.empty())
+    return Result("No Amber script found");
+
+  for (auto& info : script_info_) {
+    info.script_content = ReadScript(info.asset_name);
+    if (info.script_content.empty())
+      return Result(info.asset_name + ":\n\tEmpty Amber script");
+  }
+
+  for (auto& info : script_info_) {
+    auto shader_names = GetShaderNamesForAmberScript(info.asset_name);
+    for (const auto& shader : shader_names) {
+      auto shader_content = ReadAssetContentAsVectorUint32(shader);
+      if (shader_content.empty())
+        return Result(shader + ":\n\tEmpty shader");
+
+      info.shader_map[GetShaderID(shader)] = shader_content;
+    }
+  }
+
+  return {};
+}
+
+void AmberScriptLoader::FindAllScripts() {
+  AAssetDir* asset = AAssetManager_openDir(app_context_->activity->assetManager,
+                                           kAmberDir.c_str());
+  for (const char* file_name = AAssetDir_getNextFileName(asset); file_name;
+       file_name = AAssetDir_getNextFileName(asset)) {
+    std::string file_name_in_string(file_name);
+    if (IsEndedWith(file_name_in_string, kAmberScriptExtension)) {
+      script_info_.emplace_back();
+      script_info_.back().asset_name = file_name_in_string;
+    }
+  }
+  AAssetDir_close(asset);
+}
+
+std::vector<std::string> AmberScriptLoader::GetShaderNamesForAmberScript(
+    const std::string& script_name) {
+  std::vector<std::string> shaders;
+
+  AAssetDir* asset = AAssetManager_openDir(app_context_->activity->assetManager,
```

Comment:
Instead of reading the directory a second time, change the FindAllScripts to push all files ending in .spv into a list then this just becomes walking the list and picking out the shaders you want which should be faster then going to disk.